### PR TITLE
LPD-21792 Covering the case that a structure has more than one structureVersionId

### PR DESCRIPTION
**Pull request by vngoliveira**
[LPD-21792](https://liferay.atlassian.net/browse/LPD-21792)
[LPP-53336](https://liferay.atlassian.net/browse/LPP-53336)

### CODE REVIEW

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java [hunk 1]
```diff
@@ -52,6 +52,13 @@ protected void doUpgrade() throws Exception {
 					connection,
 					"update DDMStorageLink set structureVersionId = ? where " +
 						"storageLinkId = ?");
+			PreparedStatement preparedStatement5 = connection.prepareStatement(
+				StringBundler.concat(
+					"select distinct DDMStorageLink.structureId, (select ",
+					"max(structureVersionId) from DDMStructureVersion where ",
+					"DDMStructureVersion.structureId = ",
+					"DDMStorageLink.structureId) as structureVersionId from ",
+					"DDMStorageLink"));
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java:61]: @vngoliveira Is this `Select` the same from the previous code before [this change](https://github.com/liferay-forms/liferay-portal/commit/bc3de249c29b4b40649ceae447865efe4be4dae1#diff-847f87055f8db9aabab04cfad81ef2cdc77bb565a7a12634d449b4dc3e9bc726L24-L32)?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java [hunk 2]
```diff
@@ -104,6 +119,7 @@ protected void doUpgrade() throws Exception {
 			}
 
 			preparedStatement4.executeBatch();
+			preparedStatement3.executeBatch();
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java:122]: @vngoliveira please follow the alphabetical order, `preparedStatement3` before `preparedStatement4`

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java (hunk 1 above)
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java:61]: @vngoliveira please move all selects to be before the updates:

```
try (PreparedStatement preparedStatement1 = connection.prepareStatement(
		StringBundler.concat(
			"select DDMStorageLink.storageLinkId, ",
			"DDMStorageLink.structureId, DDMStorageLink.classPK, ",
			"DDMContent.data_ from DDMStorageLink, DDMContent where ",
			"DDMStorageLink.classPK = DDMContent.contentId"));
	PreparedStatement preparedStatement2 = connection.prepareStatement(
		"select DDMStructureVersion.structureVersionId, " +
			"DDMStructureVersion.definition from DDMStructureVersion " +
				"where structureId = ?");
	PreparedStatement preparedStatement3 = connection.prepareStatement(
		StringBundler.concat(
			"select distinct DDMStorageLink.structureId, (select ",
			"max(structureVersionId) from DDMStructureVersion where ",
			"DDMStructureVersion.structureId = ",
			"DDMStorageLink.structureId) as structureVersionId from ",
			"DDMStorageLink"));
	PreparedStatement preparedStatement4 =
		AutoBatchPreparedStatementUtil.concurrentAutoBatch(
			connection,
			"update DDMStorageLink set structureVersionId = ? where " +
				"structureId = ? and structureVersionId is null");
	PreparedStatement preparedStatement5 =
		AutoBatchPreparedStatementUtil.concurrentAutoBatch(
			connection,
			"update DDMStorageLink set structureVersionId = ? where " +
				"storageLinkId = ?");
	ResultSet resultSet1 = preparedStatement1.executeQuery()) {

	Map<Long, Long> ddmStructureVersionIds = new HashMap<>();

	while (resultSet1.next()) {
		Set<String> ddmContentFieldNames = _getFieldNames(
			resultSet1.getString("data_"), "fieldValues");

		if (SetUtil.isEmpty(ddmContentFieldNames)) {
			continue;
		}

		preparedStatement2.setLong(
			1, resultSet1.getLong("structureId"));

		try (ResultSet resultSet2 = preparedStatement2.executeQuery()) {
			while (resultSet2.next()) {
				Set<String> ddmStructureFieldNames = _getFieldNames(
					resultSet2.getString("definition"), "fields");

				if (!ddmStructureFieldNames.containsAll(
						ddmContentFieldNames)) {

					continue;
				}

				preparedStatement4.setLong(
					1, resultSet2.getLong("structureVersionId"));
				preparedStatement4.setLong(
					2, resultSet1.getLong("structureId"));

				preparedStatement4.addBatch();

				ddmStructureVersionIds.put(
					resultSet1.getLong("storageLinkId"),
					resultSet2.getLong("structureVersionId"));
			}
		}
	}

	try (ResultSet resultSet3 = preparedStatement3.executeQuery()) {
		while (resultSet3.next()) {
			preparedStatement4.setLong(
				1, resultSet3.getLong("structureVersionId"));
			preparedStatement4.setLong(
				2, resultSet3.getLong("structureId"));
			preparedStatement4.addBatch();
		}
	}

	for (Map.Entry<Long, Long> entry :
			ddmStructureVersionIds.entrySet()) {

		preparedStatement5.setLong(1, entry.getValue());
		preparedStatement5.setLong(2, entry.getKey());

		preparedStatement5.addBatch();
	}

	preparedStatement4.executeBatch();
	preparedStatement5.executeBatch();
}
```

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java:1]: @vngoliveira do we still need this code:

```
PreparedStatement preparedStatement5 =
	AutoBatchPreparedStatementUtil.concurrentAutoBatch(
		connection,
		"update DDMStorageLink set structureVersionId = ? where " +
			"storageLinkId = ?");

...

for (Map.Entry<Long, Long> entry :
		ddmStructureVersionIds.entrySet()) {

	preparedStatement5.setLong(1, entry.getValue());
	preparedStatement5.setLong(2, entry.getKey());

	preparedStatement5.addBatch();
}
```
?

This code is quite confusing so please make sure to keep only what is necessary.

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java (hunk 1 above)
Comment (vngoliveira) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java:61]: Yes, it is

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java
Comment (vngoliveira) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_1_3/DDMStorageLinkUpgradeProcess.java:1]: This code is to fix this [issue](https://liferay.atlassian.net/browse/LPD-18203), but it is not enough to fix my case reason why I needed to return the previous code before [this change](https://github.com/liferay-forms/liferay-portal/commit/bc3de249c29b4b40649ceae447865efe4be4dae1#diff-847f87055f8db9aabab04cfad81ef2cdc77bb565a7a12634d449b4dc3e9bc726L24-L32)

**Comment by vngoliveira** [comment]
ci:test:sf

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:forward

**Comment by carolmariaabb** [comment]
ci:test:stable

**Comment by carolmariaabb** [comment]
ci:forward:force

**Pull request closed without merging**